### PR TITLE
New manip modify invocation effective timeout

### PR DESCRIPTION
**Pull request by pstaeck**
Purpose of the pull request is to add a manipulation that triggers the device to set @InvocationEffectiveTimeout less than or equal to the requested threshold.

# Checklist

The following aspects have been respected by the author of this pull request, confirmed by both pull request assignee **and** reviewer:

* Changelog update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer
* README update (necessity checked and entry added or not added respectively)
  * [x] Pull Request Assignee
  * [x] Reviewer

### CODE REVIEW

File: src/t2iapi/operation/service.proto [hunk 1]
```diff
@@ -29,4 +30,11 @@ service OperationService {
     rpc SetOperatingMode (t2iapi.operation.SetOperatingModeRequest)
         returns (BasicResponse);
 
+    /*
+    Update the @InvocationEffectiveTimeout of the pm:AbstractOperationDescriptor with the given handle.
```
Comment (jannast) [inline src/t2iapi/operation/service.proto:34]: use the same wording. Update should be Modify in the description of the rpc.

**Review by jannast — CHANGES_REQUESTED** [review]

File: src/t2iapi/operation/service.proto (hunk 1 above)
Comment (pstaeck) [reply src/t2iapi/operation/service.proto]: done

Comment (pstaeck) [reply src/t2iapi/operation/service.proto]: after discussion i also added missing "an SDC operation invocation"

File: src/t2iapi/operation/operation_requests.proto [hunk 2]
```diff
@@ -20,4 +20,13 @@ Request to change the devices state, which results in setting the requested Oper
...
     OperatingMode operating_mode = 2;
+}
+
+/*
+Request to modify the devices descriptor, which results in setting the requested the @InvocationEffectiveTimeout for
+the pm:AbstractOperationDescriptor with the given handle.
+*/
+message ModifyInvocationEffectiveTimeoutRequest {
+    string handle = 1;
+    string invocation_effective_timeout = 2;
```
Comment (jannast) [inline src/t2iapi/operation/operation_requests.proto:31]: add a description to the parameters in more detail.

File: src/t2iapi/operation/service.proto [hunk 3]
```diff
@@ -29,4 +29,11 @@ service OperationService {
     rpc SetOperatingMode (t2iapi.operation.SetOperatingModeRequest)
         returns (BasicResponse);
 
+    /*
+    Modify the pm:AbstractOperationDescriptor with the given handle such that the @InvocationEffectiveTimeout is set to
+    the requested value.
+    The manipulated descriptor shall be persistent until a next manipulation call or an SDC operation invocation. If
+    the device is not able to maintain the static descriptor, it shall return RESULT_NOT_SUPPORTED.
+    */
+    rpc ModifyInvocationEffectiveTimeout (ModifyInvocationEffectiveTimeoutRequest) returns (BasicResponse);
```
Comment (jannast) [inline src/t2iapi/operation/service.proto:38]: after changing the interface, maybe rename to SetInvocationEffectiveTimeout ??

**Review by jannast — CHANGES_REQUESTED** [review]

File: src/t2iapi/operation/service.proto (hunk 3 above)
Comment (pstaeck) [reply src/t2iapi/operation/service.proto]: done - rpc is reworked

File: src/t2iapi/operation/service.proto [hunk 4]
```diff
@@ -29,4 +29,12 @@ service OperationService {
     rpc SetOperatingMode (t2iapi.operation.SetOperatingModeRequest)
         returns (BasicResponse);
 
+    /*
+    Set the @InvocationEffectiveTimeout of the pm:AbstractOperationDescriptor with the given handle to a value that is
+    less than or equal to the requested threshold.
+    The manipulated descriptor shall be persistent until a next manipulation call or an SDC operation invocation. If
+    the device is not able to maintain the static descriptor, it shall return RESULT_NOT_SUPPORTED.
+    */
+    rpc SetInvocationEffectiveTimeoutLteToThreshold (SetInvocationEffectiveTimeoutLteToThresholdRequest)
```
Comment (jannast) [inline src/t2iapi/operation/service.proto:38]: The name of the rpc seems strange SetInvocationEffectiveTimeoutLteToThreshold what means Lte??

Comment (pstaeck) [reply src/t2iapi/operation/service.proto]: yeah this was a suggestion to shorten the name of the manipulation
lte is the abbreviation of LessThanOrEqual

but your are right, it leaves room for misinterpretation, i have written it in the full length, now

**Review by jannast — APPROVED** [review]

File: src/t2iapi/operation/operation_requests.proto [hunk 5]
```diff
@@ -20,4 +21,16 @@ Request to change the devices state, which results in setting the requested Oper
...
+}
+
+/*
+Request to set the @InvocationEffectiveTimeout of the pm:AbstractOperationDescriptor with the given handle to a value
+that is less than or equal to the requested threshold.
+*/
+message SetInvocationEffectiveTimeoutLessThanOrEqualToThresholdRequest {
+    string handle = 1;  // handle of the pm:AbstractOperationDescriptor for which the @InvocationEffectiveTimeout
+                        // shall be set
+    google.protobuf.Duration threshold = 2;  // threshold value in seconds at nanosecond resolution for which applies
```
Comment (ldeichmann) [inline src/t2iapi/operation/operation_requests.proto:33]: What does the "in seconds at nanosecond resolution" mean? Do you only use the nanosecond part of the duration? That's kinda.. odd.

**Review by ldeichmann — APPROVED** [review]

**Review by jannast — APPROVED** [review]

**Review by a-kleinf — APPROVED** [review]

**Pull request merged**